# 4.8 Grouping Data & Aggregation Variables

### This notebook contains the following points:

## 1.0 Preparational Steps
## 2.0 Tasks on Exercise 4.8

## 1.0 Preparational Steps

In [1]:
# Importing libraries
import pandas as pd
import numpy as np
import os

In [2]:
# defining path
path = r'C:\Users\chris\OneDrive\Desktop\Data Analytics CF\202203_Instacart Basket Analysis'

In [3]:
# importing dataframe
ords_prods_merge = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data','ords_prods_merged_new_variables.pkl'))

In [4]:
ords_prods_merge.shape

(32404859, 19)

In [5]:
ords_prods_merge.head(10)

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2539329        1             1                   2                  8   
2   2539329        1             1                   2                  8   
3   2539329        1             1                   2                  8   
4   2539329        1             1                   2                  8   
5   2398795        1             2                   3                  7   
6   2398795        1             2                   3                  7   
7   2398795        1             2                   3                  7   
8   2398795        1             2                   3                  7   
9   2398795        1             2                   3                  7   

   days_since_prior_order  first_order  product_id  add_to_cart_order  \
0                     NaN         True         196                  1   
1                     NaN         True       14084                  2   
2                     NaN         True       12427                  3   
3                     NaN         True       26088                  4   
4                     NaN         True       26405                  5   
5                    15.0        False         196                  1   
6                    15.0        False       10258                  2   
7                    15.0        False       12427                  3   
8                    15.0        False       13176                  4   
9                    15.0        False       26088                  5   

   reordered                             product_name  aisle_id  \
0          0                                     Soda      77.0   
1          0  Organic Unsweetened Vanilla Almond Milk      91.0   
2          0                      Original Beef Jerky      23.0   
3          0               Aged White Cheddar Popcorn      23.0   
4          0         XL Pick-A-Size Paper Towel Rolls      54.0   
5          1                                     Soda      77.0   
6          0                               Pistachios     117.0   
7          1                      Original Beef Jerky      23.0   
8          0                   Bag of Organic Bananas      24.0   
9          1               Aged White Cheddar Popcorn      23.0   

   department_id  prices _merge    price_range_loc     busiest_day  \
0            7.0     9.0   both  Mid-range product  Regularly busy   
1           16.0    12.5   both  Mid-range product  Regularly busy   
2           19.0     4.4   both  Low-range product  Regularly busy   
3           19.0     4.7   both  Low-range product  Regularly busy   
4           17.0     1.0   both  Low-range product  Regularly busy   
5            7.0     9.0   both  Mid-range product  Regularly busy   
6           19.0     3.0   both  Low-range product  Regularly busy   
7           19.0     4.4   both  Low-range product  Regularly busy   
8            4.0    10.3   both  Mid-range product  Regularly busy   
9           19.0     4.7   both  Low-range product  Regularly busy   

   busiest_days busiest_period_of_day  
0  regular days        Average orders  
1  regular days        Average orders  
2  regular days        Average orders  
3  regular days        Average orders  
4  regular days        Average orders  
5  slowest days        Average orders  
6  slowest days        Average orders  
7  slowest days        Average orders  
8  slowest days        Average orders  
9  slowest days        Average orders

In [6]:
ords_prods_merge.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 32404859 entries, 0 to 32435058
Data columns (total 19 columns):
 #   Column                  Dtype   
---  ------                  -----   
 0   order_id                int64   
 1   user_id                 int64   
 2   order_number            int64   
 3   orders_day_of_week      int64   
 4   order_hour_of_day       int64   
 5   days_since_prior_order  float64 
 6   first_order             bool    
 7   product_id              int64   
 8   add_to_cart_order       int64   
 9   reordered               int64   
 10  product_name            object  
 11  aisle_id                float64 
 12  department_id           float64 
 13  prices                  float64 
 14  _merge                  category
 15  price_range_loc         object  
 16  busiest_day             object  
 17  busiest_days            object  
 18  busiest_period_of_day   object  
dtypes: bool(1), category(1), float64(4), int64(8), object(5)
memory usage: 4.4+ GB


In [8]:
# Changing the column Datatypes to save memory
ords_prods_merge['order_id'] = ords_prods_merge['order_id'].astype('int32')
ords_prods_merge['user_id'] =ords_prods_merge['user_id'].astype('int32')

In [9]:
ords_prods_merge['order_number'] =ords_prods_merge['order_number'].astype('int8')
ords_prods_merge['orders_day_of_week'] =ords_prods_merge['orders_day_of_week'].astype('int8')

In [10]:
ords_prods_merge['order_hour_of_day'] =ords_prods_merge['order_hour_of_day'].astype('int8')
ords_prods_merge['days_since_prior_order'] =ords_prods_merge['days_since_prior_order'].astype('float16')
ords_prods_merge['product_id'] =ords_prods_merge['product_id'].astype('int32')

In [11]:
ords_prods_merge['add_to_cart_order'] =ords_prods_merge['add_to_cart_order'].astype('int32')
ords_prods_merge['reordered'] =ords_prods_merge['reordered'].astype('int8')
ords_prods_merge['aisle_id'] =ords_prods_merge['aisle_id'].astype('int8')
ords_prods_merge['department_id'] =ords_prods_merge['department_id'].astype('int8')
ords_prods_merge['prices'] =ords_prods_merge['prices'].astype('float32')

In [13]:
#Dropping the '_merge' column to save memory
ords_prods_merge.drop('_merge', axis = 1, inplace = True)

In [14]:
ords_prods_merge.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 32404859 entries, 0 to 32435058
Data columns (total 18 columns):
 #   Column                  Dtype  
---  ------                  -----  
 0   order_id                int32  
 1   user_id                 int32  
 2   order_number            int8   
 3   orders_day_of_week      int8   
 4   order_hour_of_day       int8   
 5   days_since_prior_order  float16
 6   first_order             bool   
 7   product_id              int32  
 8   add_to_cart_order       int32  
 9   reordered               int8   
 10  product_name            object 
 11  aisle_id                int8   
 12  department_id           int8   
 13  prices                  float32
 14  price_range_loc         object 
 15  busiest_day             object 
 16  busiest_days            object 
 17  busiest_period_of_day   object 
dtypes: bool(1), float16(1), float32(1), int32(4), int8(6), object(5)
memory usage: 2.3+ GB


## 2.0 Tasks Exercise 4.8

2. In this Exercise, you learned how to find the aggregated mean of the “order_number” column grouped by “department_id” for a subset of your dataframe. Now, repeat this process for the entire dataframe.

In [16]:
ords_prods_merge.groupby('department_id')['order_number'].mean()

department_id
1     15.457838
2     17.277920
3     17.170395
4     17.811403
5     15.215751
6     16.439806
7     17.225802
8     15.340650
9     15.895474
10    20.197148
11    16.170638
12    15.887671
13    16.583536
14    16.773669
15    16.165037
16    17.665606
17    15.694469
18    19.310397
19    17.177343
20    16.473447
21    22.902379
Name: order_number, dtype: float64

3. Analyze the result. How do the results for the entire dataframe differ from those of the subset? Include your comments in a markdown cell below the executed code.

The results for the entire dataframe doesn't differ too much from the subset in the created along the course material. It still includes all department_ids, the values for the mean differ by max +/- 1 per department id.

4. Follow the instructions in the Exercise for creating a loyalty flag for existing customers using the transform() and loc() functions.

In [17]:
# Creating a new column for max orders using the transform function
ords_prods_merge['max_order'] = ords_prods_merge.groupby(['user_id'])['order_number'].transform(np.max)

In [18]:
ords_prods_merge.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2539329        1             1                   2                  8   
2   2539329        1             1                   2                  8   
3   2539329        1             1                   2                  8   
4   2539329        1             1                   2                  8   

   days_since_prior_order  first_order  product_id  add_to_cart_order  \
0                     NaN         True         196                  1   
1                     NaN         True       14084                  2   
2                     NaN         True       12427                  3   
3                     NaN         True       26088                  4   
4                     NaN         True       26405                  5   

   reordered                             product_name  aisle_id  \
0          0                                     Soda        77   
1          0  Organic Unsweetened Vanilla Almond Milk        91   
2          0                      Original Beef Jerky        23   
3          0               Aged White Cheddar Popcorn        23   
4          0         XL Pick-A-Size Paper Towel Rolls        54   

   department_id  prices    price_range_loc     busiest_day  busiest_days  \
0              7     9.0  Mid-range product  Regularly busy  regular days   
1             16    12.5  Mid-range product  Regularly busy  regular days   
2             19     4.4  Low-range product  Regularly busy  regular days   
3             19     4.7  Low-range product  Regularly busy  regular days   
4             17     1.0  Low-range product  Regularly busy  regular days   

  busiest_period_of_day  max_order  
0        Average orders         10  
1        Average orders         10  
2        Average orders         10  
3        Average orders         10  
4        Average orders         10

In [19]:
# Creating a new column loyalty_flag using the loc function
ords_prods_merge.loc[ords_prods_merge['max_order'] > 40, 'loyalty_flag'] = 'Loyal customer'

In [20]:
ords_prods_merge.loc[(ords_prods_merge['max_order'] <= 40) & (ords_prods_merge['max_order'] > 10), 'loyalty_flag'] = 'Regular customer'

In [21]:
ords_prods_merge.loc[ords_prods_merge['max_order'] <= 10, 'loyalty_flag'] = 'New customer'

In [22]:
ords_prods_merge['loyalty_flag'].value_counts(dropna = False)

Regular customer    15876776
Loyal customer      10284093
New customer         6243990
Name: loyalty_flag, dtype: int64

5. The marketing team at Instacart wants to know whether there’s a difference between the spending habits of the three types of customers you identified. Use the loyalty flag you created and check the basic statistics of the product prices for each loyalty category (Loyal Customer, Regular Customer, and New Customer). What you’re trying to determine is whether the prices of products purchased by loyal customers differ from those purchased by regular or new customers.

In [23]:
# Grouping the df by loyalty_flag column and using describe to get the basic statistics for price column per loyalty_flag
ords_prods_merge.groupby('loyalty_flag')['prices'].describe()

count       mean         std  min  25%  50%   75%  \
loyalty_flag                                                               
Loyal customer    10284093.0  10.407317  327.981293  1.0  4.2  7.4  11.2   
New customer       6243990.0  13.302089  597.521545  1.0  4.2  7.4  11.3   
Regular customer  15876776.0  12.389329  539.679504  1.0  4.2  7.4  11.3   

                      max  
loyalty_flag               
Loyal customer    99999.0  
New customer      99999.0  
Regular customer  99999.0

The graph shows, that loyal customers have the lowest mean with ~ 10.4 spent as average price, while the new customers have the highest average with 13.3. The regular customers have an average of ~12.5. Looking at the Quartiles you can see that the spending is distributed very evenly for the three groups with the exception of 75% being just a little lower with 11.2 compared to 11.3 for new and regular customers.
Another thing that strikes the eye ist the max price spent with 99999 which is quite far off the other values (almost 30x the standard deviation) indicating an outlier.
However, what this does not show is the average basket per customer (or average total spent per order).

6. The team now wants to target different types of spenders in their marketing campaigns. This can be achieved by looking at the prices of the items people are buying. Create a spending flag for each user based on the average price across all their orders using the following criteria:

    If the mean of the prices of products purchased by a user is lower than 10, then flag them as a “Low spender.”
    If the mean of the prices of products purchased by a user is higher than or equal to 10, then flag them as a “High spender.”


In [24]:
#Testing calculation of mean for price column per user id
ords_prods_merge.groupby('user_id')['prices'].transform(np.mean)

0           6.367797
1           6.367797
2           6.367797
3           6.367797
4           6.367797
              ...   
32435054    7.058915
32435055    7.058915
32435056    7.058915
32435057    7.058915
32435058    7.058915
Name: prices, Length: 32404859, dtype: float32

In [25]:
#Creating new column average_price 
ords_prods_merge['average_price'] = ords_prods_merge.groupby('user_id')['prices'].transform(np.mean)

In [26]:
ords_prods_merge.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2539329        1             1                   2                  8   
2   2539329        1             1                   2                  8   
3   2539329        1             1                   2                  8   
4   2539329        1             1                   2                  8   

   days_since_prior_order  first_order  product_id  add_to_cart_order  \
0                     NaN         True         196                  1   
1                     NaN         True       14084                  2   
2                     NaN         True       12427                  3   
3                     NaN         True       26088                  4   
4                     NaN         True       26405                  5   

   reordered  ... aisle_id  department_id  prices    price_range_loc  \
0          0  ...       77              7     9.0  Mid-range product   
1          0  ...       91             16    12.5  Mid-range product   
2          0  ...       23             19     4.4  Low-range product   
3          0  ...       23             19     4.7  Low-range product   
4          0  ...       54             17     1.0  Low-range product   

      busiest_day  busiest_days busiest_period_of_day max_order  loyalty_flag  \
0  Regularly busy  regular days        Average orders        10  New customer   
1  Regularly busy  regular days        Average orders        10  New customer   
2  Regularly busy  regular days        Average orders        10  New customer   
3  Regularly busy  regular days        Average orders        10  New customer   
4  Regularly busy  regular days        Average orders        10  New customer   

  average_price  
0      6.367797  
1      6.367797  
2      6.367797  
3      6.367797  
4      6.367797  

[5 rows x 21 columns]

In [27]:
#Creating new column spending flag for high (greater and equal to 10) and low (lower 10) spenders
ords_prods_merge.loc[ords_prods_merge['average_price'] >= 10, 'spending_flag'] = 'High spender'

In [28]:
ords_prods_merge.loc[ords_prods_merge['average_price'] < 10, 'spending_flag'] = 'Low spender'

In [29]:
ords_prods_merge.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2539329        1             1                   2                  8   
2   2539329        1             1                   2                  8   
3   2539329        1             1                   2                  8   
4   2539329        1             1                   2                  8   

   days_since_prior_order  first_order  product_id  add_to_cart_order  \
0                     NaN         True         196                  1   
1                     NaN         True       14084                  2   
2                     NaN         True       12427                  3   
3                     NaN         True       26088                  4   
4                     NaN         True       26405                  5   

   reordered  ... department_id  prices    price_range_loc     busiest_day  \
0          0  ...             7     9.0  Mid-range product  Regularly busy   
1          0  ...            16    12.5  Mid-range product  Regularly busy   
2          0  ...            19     4.4  Low-range product  Regularly busy   
3          0  ...            19     4.7  Low-range product  Regularly busy   
4          0  ...            17     1.0  Low-range product  Regularly busy   

   busiest_days busiest_period_of_day max_order  loyalty_flag  average_price  \
0  regular days        Average orders        10  New customer       6.367797   
1  regular days        Average orders        10  New customer       6.367797   
2  regular days        Average orders        10  New customer       6.367797   
3  regular days        Average orders        10  New customer       6.367797   
4  regular days        Average orders        10  New customer       6.367797   

  spending_flag  
0   Low spender  
1   Low spender  
2   Low spender  
3   Low spender  
4   Low spender  

[5 rows x 22 columns]

In [30]:
ords_prods_merge.tail()

order_id  user_id  order_number  orders_day_of_week  \
32435054   2977660   206209            13                   1   
32435055   2977660   206209            13                   1   
32435056   2977660   206209            13                   1   
32435057   2977660   206209            13                   1   
32435058   2977660   206209            13                   1   

          order_hour_of_day  days_since_prior_order  first_order  product_id  \
32435054                 12                     7.0        False       14197   
32435055                 12                     7.0        False       38730   
32435056                 12                     7.0        False       31477   
32435057                 12                     7.0        False        6567   
32435058                 12                     7.0        False       22920   

          add_to_cart_order  reordered  ... department_id  prices  \
32435054                  5          1  ...             9     5.6   
32435055                  6          0  ...            19     5.9   
32435056                  7          0  ...            19     4.2   
32435057                  8          0  ...            19     4.9   
32435058                  9          0  ...            19     1.7   

            price_range_loc     busiest_day  busiest_days  \
32435054  Mid-range product  Regularly busy  busiest days   
32435055  Mid-range product  Regularly busy  busiest days   
32435056  Low-range product  Regularly busy  busiest days   
32435057  Low-range product  Regularly busy  busiest days   
32435058  Low-range product  Regularly busy  busiest days   

         busiest_period_of_day max_order      loyalty_flag  average_price  \
32435054           Most orders        13  Regular customer       7.058915   
32435055           Most orders        13  Regular customer       7.058915   
32435056           Most orders        13  Regular customer       7.058915   
32435057           Most orders        13  Regular customer       7.058915   
32435058           Most orders        13  Regular customer       7.058915   

         spending_flag  
32435054   Low spender  
32435055   Low spender  
32435056   Low spender  
32435057   Low spender  
32435058   Low spender  

[5 rows x 22 columns]

In [32]:
ords_prods_merge['spending_flag'].value_counts(dropna = False)

Low spender     31770614
High spender      634245
Name: spending_flag, dtype: int64

7. In order to send relevant notifications to users within the app (for instance, asking users if they want to buy the same item again), the Instacart team wants you to determine frequent versus non-frequent customers. Create an order frequency flag that marks the regularity of a user’s ordering behavior according to the median in the “days_since_prior_order” column. The criteria for the flag should be as follows:

        If the median of “days_since_prior_order” is higher than 20, then the customer should be labeled a “Non-frequent customer.”
    
        If the median is higher than 10 and lower than or equal to 20, then the customer should be labeled a “Regular customer.”
    
        If the median is lower than or equal to 10, then the customer should be labeled a “Frequent customer.”


### Creating a test dataframe to try out procedure on small dataframe before continuing with task 7

In [33]:
# Creating a test dataframe for user id 1
df_testuser = ords_prods_merge.loc[ords_prods_merge['user_id'] == 1]

In [34]:
df_testuser

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0    2539329        1             1                   2                  8   
1    2539329        1             1                   2                  8   
2    2539329        1             1                   2                  8   
3    2539329        1             1                   2                  8   
4    2539329        1             1                   2                  8   
5    2398795        1             2                   3                  7   
6    2398795        1             2                   3                  7   
7    2398795        1             2                   3                  7   
8    2398795        1             2                   3                  7   
9    2398795        1             2                   3                  7   
10   2398795        1             2                   3                  7   
11    473747        1             3                   3                 12   
12    473747        1             3                   3                 12   
13    473747        1             3                   3                 12   
14    473747        1             3                   3                 12   
15    473747        1             3                   3                 12   
16   2254736        1             4                   4                  7   
17   2254736        1             4                   4                  7   
18   2254736        1             4                   4                  7   
19   2254736        1             4                   4                  7   
20   2254736        1             4                   4                  7   
21    431534        1             5                   4                 15   
22    431534        1             5                   4                 15   
23    431534        1             5                   4                 15   
24    431534        1             5                   4                 15   
25    431534        1             5                   4                 15   
26    431534        1             5                   4                 15   
27    431534        1             5                   4                 15   
28    431534        1             5                   4                 15   
29   3367565        1             6                   2                  7   
30   3367565        1             6                   2                  7   
31   3367565        1             6                   2                  7   
32   3367565        1             6                   2                  7   
33    550135        1             7                   1                  9   
34    550135        1             7                   1                  9   
35    550135        1             7                   1                  9   
36    550135        1             7                   1                  9   
37    550135        1             7                   1                  9   
38   3108588        1             8                   1                 14   
39   3108588        1             8                   1                 14   
40   3108588        1             8                   1                 14   
41   3108588        1             8                   1                 14   
42   3108588        1             8                   1                 14   
43   3108588        1             8                   1                 14   
44   2295261        1             9                   1                 16   
45   2295261        1             9                   1                 16   
46   2295261        1             9                   1                 16   
47   2295261        1             9                   1                 16   
48   2295261        1             9                   1                 16   
49   2295261        1             9                   1                 16   
50   2550362        1     

In [35]:
# Identifying the median for the entire days since prior order column
df_testuser['days_since_prior_order'].median()

20.5

In [36]:
# Identifying the median for days since prior order with no duplicate order numbers
df_testuser.drop_duplicates(subset = ['order_number'])['days_since_prior_order'].median()

20.0

In [37]:
df_testuser[df_testuser['days_since_prior_order'] > 20.5]['days_since_prior_order'].value_counts()

30.0    9
28.0    8
21.0    5
29.0    5
Name: days_since_prior_order, dtype: int64

In [38]:
df_testuser[df_testuser['days_since_prior_order'] < 20.5]['days_since_prior_order'].value_counts()

15.0    6
14.0    6
0.0     6
20.0    5
19.0    4
Name: days_since_prior_order, dtype: int64

Correct Median for user id 1 is 20, so I will use the drop_duplicates to proceed and identify the median for days since prior order per customer

In [39]:
# Testing the identified example of dropping duplicates for all users
ords_prods_merge.drop_duplicates(subset = ['user_id', 'order_number']).groupby('user_id')['days_since_prior_order'].median()

user_id
1         20.0
2         13.0
3         11.0
4         17.0
5         11.0
          ... 
206205    20.0
206206     3.0
206207     8.0
206208     7.0
206209    20.0
Name: days_since_prior_order, Length: 206209, dtype: float16

In [40]:
# Defining Median values as a dataframe so it can be merged in the next step
df_median_values = ords_prods_merge.drop_duplicates(subset = ['user_id', 'order_number']).groupby('user_id')['days_since_prior_order'].median()

In [41]:
# Testing merge on testuser dataframe
df_testuser.merge(df_median_values, on = ['user_id'], how = 'left')

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0    2539329        1             1                   2                  8   
1    2539329        1             1                   2                  8   
2    2539329        1             1                   2                  8   
3    2539329        1             1                   2                  8   
4    2539329        1             1                   2                  8   
5    2398795        1             2                   3                  7   
6    2398795        1             2                   3                  7   
7    2398795        1             2                   3                  7   
8    2398795        1             2                   3                  7   
9    2398795        1             2                   3                  7   
10   2398795        1             2                   3                  7   
11    473747        1             3                   3                 12   
12    473747        1             3                   3                 12   
13    473747        1             3                   3                 12   
14    473747        1             3                   3                 12   
15    473747        1             3                   3                 12   
16   2254736        1             4                   4                  7   
17   2254736        1             4                   4                  7   
18   2254736        1             4                   4                  7   
19   2254736        1             4                   4                  7   
20   2254736        1             4                   4                  7   
21    431534        1             5                   4                 15   
22    431534        1             5                   4                 15   
23    431534        1             5                   4                 15   
24    431534        1             5                   4                 15   
25    431534        1             5                   4                 15   
26    431534        1             5                   4                 15   
27    431534        1             5                   4                 15   
28    431534        1             5                   4                 15   
29   3367565        1             6                   2                  7   
30   3367565        1             6                   2                  7   
31   3367565        1             6                   2                  7   
32   3367565        1             6                   2                  7   
33    550135        1             7                   1                  9   
34    550135        1             7                   1                  9   
35    550135        1             7                   1                  9   
36    550135        1             7                   1                  9   
37    550135        1             7                   1                  9   
38   3108588        1             8                   1                 14   
39   3108588        1             8                   1                 14   
40   3108588        1             8                   1                 14   
41   3108588        1             8                   1                 14   
42   3108588        1             8                   1                 14   
43   3108588        1             8                   1                 14   
44   2295261        1             9                   1                 16   
45   2295261        1             9                   1                 16   
46   2295261        1             9                   1                 16   
47   2295261        1             9                   1                 16   
48   2295261        1             9                   1                 16   
49   2295261        1             9                   1                 16   
50   2550362        1     

In [42]:
# Performing merge on testuser dataframe
df_testuser_merge = df_testuser.merge(df_median_values, on = ['user_id'], how = 'left')

In [43]:
df_testuser_merge

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0    2539329        1             1                   2                  8   
1    2539329        1             1                   2                  8   
2    2539329        1             1                   2                  8   
3    2539329        1             1                   2                  8   
4    2539329        1             1                   2                  8   
5    2398795        1             2                   3                  7   
6    2398795        1             2                   3                  7   
7    2398795        1             2                   3                  7   
8    2398795        1             2                   3                  7   
9    2398795        1             2                   3                  7   
10   2398795        1             2                   3                  7   
11    473747        1             3                   3                 12   
12    473747        1             3                   3                 12   
13    473747        1             3                   3                 12   
14    473747        1             3                   3                 12   
15    473747        1             3                   3                 12   
16   2254736        1             4                   4                  7   
17   2254736        1             4                   4                  7   
18   2254736        1             4                   4                  7   
19   2254736        1             4                   4                  7   
20   2254736        1             4                   4                  7   
21    431534        1             5                   4                 15   
22    431534        1             5                   4                 15   
23    431534        1             5                   4                 15   
24    431534        1             5                   4                 15   
25    431534        1             5                   4                 15   
26    431534        1             5                   4                 15   
27    431534        1             5                   4                 15   
28    431534        1             5                   4                 15   
29   3367565        1             6                   2                  7   
30   3367565        1             6                   2                  7   
31   3367565        1             6                   2                  7   
32   3367565        1             6                   2                  7   
33    550135        1             7                   1                  9   
34    550135        1             7                   1                  9   
35    550135        1             7                   1                  9   
36    550135        1             7                   1                  9   
37    550135        1             7                   1                  9   
38   3108588        1             8                   1                 14   
39   3108588        1             8                   1                 14   
40   3108588        1             8                   1                 14   
41   3108588        1             8                   1                 14   
42   3108588        1             8                   1                 14   
43   3108588        1             8                   1                 14   
44   2295261        1             9                   1                 16   
45   2295261        1             9                   1                 16   
46   2295261        1             9                   1                 16   
47   2295261        1             9                   1                 16   
48   2295261        1             9                   1                 16   
49   2295261        1             9                   1                 16   
50   2550362        1     

In [44]:
# Renaming column to better match column values
df_testuser_merge.rename(columns = {'days_since_prior_order_y':'median_days_since_prior_order'})

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0    2539329        1             1                   2                  8   
1    2539329        1             1                   2                  8   
2    2539329        1             1                   2                  8   
3    2539329        1             1                   2                  8   
4    2539329        1             1                   2                  8   
5    2398795        1             2                   3                  7   
6    2398795        1             2                   3                  7   
7    2398795        1             2                   3                  7   
8    2398795        1             2                   3                  7   
9    2398795        1             2                   3                  7   
10   2398795        1             2                   3                  7   
11    473747        1             3                   3                 12   
12    473747        1             3                   3                 12   
13    473747        1             3                   3                 12   
14    473747        1             3                   3                 12   
15    473747        1             3                   3                 12   
16   2254736        1             4                   4                  7   
17   2254736        1             4                   4                  7   
18   2254736        1             4                   4                  7   
19   2254736        1             4                   4                  7   
20   2254736        1             4                   4                  7   
21    431534        1             5                   4                 15   
22    431534        1             5                   4                 15   
23    431534        1             5                   4                 15   
24    431534        1             5                   4                 15   
25    431534        1             5                   4                 15   
26    431534        1             5                   4                 15   
27    431534        1             5                   4                 15   
28    431534        1             5                   4                 15   
29   3367565        1             6                   2                  7   
30   3367565        1             6                   2                  7   
31   3367565        1             6                   2                  7   
32   3367565        1             6                   2                  7   
33    550135        1             7                   1                  9   
34    550135        1             7                   1                  9   
35    550135        1             7                   1                  9   
36    550135        1             7                   1                  9   
37    550135        1             7                   1                  9   
38   3108588        1             8                   1                 14   
39   3108588        1             8                   1                 14   
40   3108588        1             8                   1                 14   
41   3108588        1             8                   1                 14   
42   3108588        1             8                   1                 14   
43   3108588        1             8                   1                 14   
44   2295261        1             9                   1                 16   
45   2295261        1             9                   1                 16   
46   2295261        1             9                   1                 16   
47   2295261        1             9                   1                 16   
48   2295261        1             9                   1                 16   
49   2295261        1             9                   1                 16   
50   2550362        1     

### Applying results from testuser dataframe to entire dataframe

In [45]:
# Applying test on entire dataframe
ords_prods_merge = ords_prods_merge.merge(df_median_values, on = ['user_id'], how = 'left')

In [46]:
ords_prods_merge.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2539329        1             1                   2                  8   
2   2539329        1             1                   2                  8   
3   2539329        1             1                   2                  8   
4   2539329        1             1                   2                  8   

   days_since_prior_order_x  first_order  product_id  add_to_cart_order  \
0                       NaN         True         196                  1   
1                       NaN         True       14084                  2   
2                       NaN         True       12427                  3   
3                       NaN         True       26088                  4   
4                       NaN         True       26405                  5   

   reordered  ... prices    price_range_loc     busiest_day  busiest_days  \
0          0  ...    9.0  Mid-range product  Regularly busy  regular days   
1          0  ...   12.5  Mid-range product  Regularly busy  regular days   
2          0  ...    4.4  Low-range product  Regularly busy  regular days   
3          0  ...    4.7  Low-range product  Regularly busy  regular days   
4          0  ...    1.0  Low-range product  Regularly busy  regular days   

  busiest_period_of_day max_order  loyalty_flag average_price  spending_flag  \
0        Average orders        10  New customer      6.367797    Low spender   
1        Average orders        10  New customer      6.367797    Low spender   
2        Average orders        10  New customer      6.367797    Low spender   
3        Average orders        10  New customer      6.367797    Low spender   
4        Average orders        10  New customer      6.367797    Low spender   

  days_since_prior_order_y  
0                     20.0  
1                     20.0  
2                     20.0  
3                     20.0  
4                     20.0  

[5 rows x 23 columns]

In [47]:
#Checking Memory usage before final steps
ords_prods_merge.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 32404859 entries, 0 to 32404858
Data columns (total 23 columns):
 #   Column                    Dtype  
---  ------                    -----  
 0   order_id                  int32  
 1   user_id                   int32  
 2   order_number              int8   
 3   orders_day_of_week        int8   
 4   order_hour_of_day         int8   
 5   days_since_prior_order_x  float16
 6   first_order               bool   
 7   product_id                int32  
 8   add_to_cart_order         int32  
 9   reordered                 int8   
 10  product_name              object 
 11  aisle_id                  int8   
 12  department_id             int8   
 13  prices                    float32
 14  price_range_loc           object 
 15  busiest_day               object 
 16  busiest_days              object 
 17  busiest_period_of_day     object 
 18  max_order                 int8   
 19  loyalty_flag              object 
 20  average_price         

In [48]:
ords_prods_merge.loc[ords_prods_merge['days_since_prior_order_y'] > 20, 'Frequency_flag'] = 'Non-frequent customer'

In [49]:
ords_prods_merge.loc[(ords_prods_merge['days_since_prior_order_y'] <= 20) & (ords_prods_merge['days_since_prior_order_y'] > 10), 'Frequency_flag'] = 'Regular customer'

In [50]:
ords_prods_merge.loc[ords_prods_merge['days_since_prior_order_y'] <= 10, 'Frequency_flag'] = 'Frequent customer'

In [51]:
ords_prods_merge.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 32404859 entries, 0 to 32404858
Data columns (total 24 columns):
 #   Column                    Dtype  
---  ------                    -----  
 0   order_id                  int32  
 1   user_id                   int32  
 2   order_number              int8   
 3   orders_day_of_week        int8   
 4   order_hour_of_day         int8   
 5   days_since_prior_order_x  float16
 6   first_order               bool   
 7   product_id                int32  
 8   add_to_cart_order         int32  
 9   reordered                 int8   
 10  product_name              object 
 11  aisle_id                  int8   
 12  department_id             int8   
 13  prices                    float32
 14  price_range_loc           object 
 15  busiest_day               object 
 16  busiest_days              object 
 17  busiest_period_of_day     object 
 18  max_order                 int8   
 19  loyalty_flag              object 
 20  average_price         

In [52]:
# Renaming column to better match column values
ords_prods_merge.rename(columns = {'days_since_prior_order_y':'median_days_since_prior_order'}, inplace = True)

In [53]:
ords_prods_merge['Frequency_flag'].value_counts(dropna = False)

Frequent customer        21903428
Regular customer          7121704
Non-frequent customer     3379722
NaN                             5
Name: Frequency_flag, dtype: int64

In [54]:
ords_prods_merge[ords_prods_merge['Frequency_flag'].isnull()]

order_id  user_id  order_number  orders_day_of_week  \
25156870    895835   159838             1                   0   
25156871    895835   159838             1                   0   
25156872    895835   159838             1                   0   
25156873    895835   159838             1                   0   
25156874    895835   159838             1                   0   

          order_hour_of_day  days_since_prior_order_x  first_order  \
25156870                 17                       NaN         True   
25156871                 17                       NaN         True   
25156872                 17                       NaN         True   
25156873                 17                       NaN         True   
25156874                 17                       NaN         True   

          product_id  add_to_cart_order  reordered  ...    price_range_loc  \
25156870       22198                  1          0  ...  Low-range product   
25156871       23695                  2          0  ...  Low-range product   
25156872       10749                  3          0  ...  Mid-range product   
25156873       21334                  5          0  ...  Mid-range product   
25156874       33401                  6          0  ...  Mid-range product   

          busiest_day  busiest_days  busiest_period_of_day max_order  \
25156870  Busiest day  busiest days         Average orders         1   
25156871  Busiest day  busiest days         Average orders         1   
25156872  Busiest day  busiest days         Average orders         1   
25156873  Busiest day  busiest days         Average orders         1   
25156874  Busiest day  busiest days         Average orders         1   

          loyalty_flag average_price spending_flag  \
25156870  New customer          7.42   Low spender   
25156871  New customer          7.42   Low spender   
25156872  New customer          7.42   Low spender   
25156873  New customer          7.42   Low spender   
25156874  New customer          7.42   Low spender   

          median_days_since_prior_order Frequency_flag  
25156870                            NaN            NaN  
25156871                            NaN            NaN  
25156872                            NaN            NaN  
25156873                            NaN            NaN  
25156874                            NaN            NaN  

[5 rows x 24 columns]

There is one user that only made 1 purchase, therefore no median frequency can be calculated.
--> The NaN values will be assigned to Non-frequent customer

In [55]:
#Changing NaN values of the Frequency flag column to non-frequent customers
ords_prods_merge['Frequency_flag'].fillna('Non-frequent customer', inplace = True)

In [56]:
ords_prods_merge['Frequency_flag'].value_counts(dropna = False)

Frequent customer        21903428
Regular customer          7121704
Non-frequent customer     3379727
Name: Frequency_flag, dtype: int64

In [57]:
#Checking dataframe row count
ords_prods_merge.shape

(32404859, 24)

In [58]:
ords_prods_merge.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 32404859 entries, 0 to 32404858
Data columns (total 24 columns):
 #   Column                         Dtype  
---  ------                         -----  
 0   order_id                       int32  
 1   user_id                        int32  
 2   order_number                   int8   
 3   orders_day_of_week             int8   
 4   order_hour_of_day              int8   
 5   days_since_prior_order_x       float16
 6   first_order                    bool   
 7   product_id                     int32  
 8   add_to_cart_order              int32  
 9   reordered                      int8   
 10  product_name                   object 
 11  aisle_id                       int8   
 12  department_id                  int8   
 13  prices                         float32
 14  price_range_loc                object 
 15  busiest_day                    object 
 16  busiest_days                   object 
 17  busiest_period_of_day          object 
 18  

In [60]:
#Exporting File to Prepared Data folder
ords_prods_merge.to_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'ords_prods_merge_grouped.pkl'))